File: main/framework/api/agents.py
from fastapi import APIRouter, HTTPException, status

router = APIRouter(prefix="/api/v1/agents", tags=["agents"])


@router.get("")
async def list_agents():
    """List all registered agents."""
    from main.framework.core.agent_registry import registry

    agents = registry.list_agents()
    return [
        {
            "name": a.name,
            "description": a.description,
            "capabilities": a.capabilities,
            "tools": a.tools,
            "mode": a.mode,
        }
        for a in agents
    ]


@router.get("/stats")
async def agent_stats():
    """Agent usage stats from workflow execution nodes."""
    try:
        from sqlalchemy import func
        from main.framework.models.database import SessionLocal
        from main.framework.models.workflow_execution import ExecutionNode
        from main.framework.core.agent_registry import registry

        db = SessionLocal()
        try:
            rows = db.query(
                ExecutionNode.agent, ExecutionNode.status, func.count(ExecutionNode.id)
            ).group_by(ExecutionNode.agent, ExecutionNode.status).all()
            stats: dict = {}
            for agent, s, count in rows:
                if agent not in stats:
                    stats[agent] = {"total": 0, "completed": 0, "failed": 0}
                stats[agent][s] = stats[agent].get(s, 0) + count
                stats[agent]["total"] += count
        finally:
            db.close()

        result = []
        for a in registry.list_agents():
            s = stats.get(a.name, {"total": 0, "completed": 0, "failed": 0})
            total_terminal = s["completed"] + s["failed"]
            result.append({
                "name": a.name, "description": a.description, "mode": a.mode,
                "executions_total": s["total"], "executions_completed": s["completed"],
                "executions_failed": s["failed"],
                "success_rate": round(s["completed"] / max(total_terminal, 1) * 100, 1),
            })
        return result
    except Exception:
        return []


@router.get("/{name}")
async def get_agent(name: str):
    """Get agent details by name."""
    from main.framework.core.agent_registry import registry

    agent = registry.get_agent(name)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")
    return {
        "name": agent.name,
        "description": agent.description,
        "capabilities": agent.capabilities,
        "tools": agent.tools,
        "mode": agent.mode,
    }

File: webui/src/api/agents.ts
/**
 * Typed wrappers for the agent registry API.
 *
 * Mirrors the three routes declared in
 * `main/framework/controllers/agents.py`:
 *
 *   GET /api/v1/agents          → list of agent summaries
 *   GET /api/v1/agents/stats    → per-agent execution telemetry
 *   GET /api/v1/agents/{name}   → single agent summary (404 on miss)
 *
 * The backend returns snake_case fields, which line up with the
 * canonical `Agent` and `AgentDetail` shapes in `../types/agent` —
 * no transformation is needed on the client.
 *
 * Notes
 * -----
 * - `getAgentStats` returns a `Record<name, …>` so callers can index
 *   by agent name without scanning an array.
 * - `getAgent` URL-encodes the name so spaces / slashes round-trip
 *   safely through the FastAPI path matcher.
 */

import { API_V1_BASE } from '../config/env';
import type { Agent, AgentDetail } from '../types/agent';

/**
 * Per-agent execution telemetry returned by `/api/v1/agents/stats`.
 *
 * Each entry contains agent name, description, mode, and execution stats.
 */
export interface AgentStatsEntry {
  /** Agent name. */
  name: string;
  /** Agent description. */
  description: string;
  /** Agent mode (agent, fusion, orchestrator). */
  mode: string;
  /** Total number of executions. */
  executions_total: number;
  /** Number of completed executions. */
  executions_completed: number;
  /** Number of failed executions. */
  executions_failed: number;
  /** Percent of executions that completed without error (0–100). */
  success_rate: number;
}

/**
 * Fetch every registered agent.
 *
 * `GET /api/v1/agents` → `Agent[]`
 */
export async function listAgents(): Promise<Agent[]> {
  const res = await fetch(`${API_V1_BASE}/agents`);
  if (!res.ok) {
    throw new Error(
      `listAgents failed: ${res.status} ${res.statusText}`,
    );
  }
  return (await res.json()) as Agent[];
}

/**
 * Fetch execution telemetry for all agents in one round-trip.
 *
 * `GET /api/v1/agents/stats` → `AgentStatsEntry[]`
 */
export async function getAgentStats(): Promise<AgentStatsEntry[]> {
  const res = await fetch(`${API_V1_BASE}/agents/stats`);
  if (!res.ok) {
    throw new Error(
      `getAgentStats failed: ${res.status} ${res.statusText}`,
    );
  }
  return (await res.json()) as AgentStatsEntry[];
}

/**
 * Fetch a single agent by registry name.
 *
 * `GET /api/v1/agents/{name}` → `AgentDetail`
 *
 * The backend raises 404 when `name` is unknown; the resulting
 * `TypeError` is rethrown so callers can branch on the status code.
 */
export async function getAgent(name: string): Promise<AgentDetail> {
  const res = await fetch(
    `${API_V1_BASE}/agents/${encodeURIComponent(name)}`,
  );
  if (!res.ok) {
    throw new Error(
      `getAgent(${name}) failed: ${res.status} ${res.statusText}`,
    );
  }
  return (await res.json()) as AgentDetail;
}
